node/src/actors/chain_manager/mod.rs

```diff
@@ -347,57 +362,93 @@ impl ChainManager {
+                let vrf_proof = match block.block_header.proof.proof.proof_to_hash(vrf_ctx) {
+                    Ok(vrf) => vrf,
+                    Err(e) => {
+                        log::warn!("Block candidate has an invalid vrf proof: {}", e);
+                        return;
+                    }
+                };
+
+                if let Some(best_candidate) = &self.best_candidate {
+                    if compare_block_candidates(
+                        hash_block,
+                        reputation,
+                        vrf_proof,
+                        best_candidate.block.hash(),
+                        best_candidate.reputation,
+                        best_candidate.vrf_proof,
+                        &target_vrf_slots,
+                    ) != Ordering::Greater
+                    {
+                        log::debug!("Ignoring new candidate due to have one better");
+                        return;
+                    }
+                }
+                match process_validations(
                     &block,
                     current_epoch,
-                    prev_block_hash,
                     vrf_input,
-                    &mut signatures_to_verify,
-                    total_identities,
-                    mining_bf,
-                )
-                .map_err(Into::into)
-                .and_then(|()| {
-                    // This only verifies the block VRF proof and returns the vrf_hash
-                    // It is tested in
-                    // calling_validate_candidate_and_then_verify_signatures_returns_block_vrf_hash
-                    verify_signatures(signatures_to_verify, vrf_ctx, secp_ctx)
-                }) {
-                    Ok(vrf_hash) => {
-                        self.candidates
-                            .insert(hash_block, (block.clone(), vrf_hash[0]));
+                    chain_info.highest_block_checkpoint,
+                    rep_engine,
+                    self.epoch_constants.unwrap(),
+                    &self.chain_state.unspent_outputs_pool,
+                    &self.chain_state.data_request_pool,
+                    // The unwrap is safe because if there is no VRF context,
+                    // the actor should have stopped execution
+                    self.vrf_ctx.as_mut().unwrap(),
+                    self.secp.as_ref().unwrap(),
+                    chain_info.consensus_constants.mining_backup_factor,
+                    chain_info.consensus_constants.bootstrap_hash,
+                    chain_info.consensus_constants.genesis_hash,
+                    self.chain_state.block_number(),
+                    chain_info.consensus_constants.collateral_minimum,
+                    chain_info.consensus_constants.collateral_age,
+                    chain_info.consensus_constants.max_vt_weight,
+                    chain_info.consensus_constants.max_dr_weight,
+                ) {
```

Comment:
Lol at these arguments. Maybe it is time to pass `chain_info.consensus_constants` as a whole? :rofl:

Reply:
I thought the same :P